pkg/gitopsjob/job.go

```diff
@@ -0,0 +1,240 @@
+									Value: gitopsConfig.Context,
+								},
+								{
+									Name:  "NAMESPACE",
+									Value: resourceNamespace,
+								},
+								{
+									Name: "GITHUB_TOKEN",
+									ValueFrom: &corev1.EnvVarSource{
+										SecretKeyRef: &corev1.SecretKeySelector{
+											LocalObjectReference: corev1.LocalObjectReference{
+												Name: "has-github-token",
+											},
+											Key: "token",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	err := client.Create(ctx, &gitopsJob)
+	return err
+}
+
+// WaitForJob waits for the given Kubernetes job to complete.
+// If it errors out, or the given timeout is reached, an error is returned
+func WaitForJob(log logr.Logger, ctx context.Context, client ctrlclient.Client, clientset kubernetes.Interface, jobName string, jobNamespace string, timeout time.Duration) error {
+	var job batchv1.Job
+	var err error
+	for stay, timeout := true, time.After(timeout); stay; {
+		err = client.Get(context.Background(), types.NamespacedName{Namespace: jobNamespace, Name: jobName}, &job)
+		if err != nil {
+			// If the error is anything but a isnotfound error, return the error
+			// If the resource wasn't found, keep trying up to the timeout, in case the job hasn't appeared yet
+			if !k8sErrors.IsNotFound(err) {
+				return err
+			}
+		}
+
+		// The CompletionTime in the job's status will only get set when the Job completes successfully, so check for its presence
+		if job.Status.CompletionTime != nil {
+			return nil
+		}
+		if job.Status.Failed >= 5 {
+			// Attempt to get the Pod logs
+			errMsg, err := GetPodLogs(context.Background(), client, clientset, jobName, jobNamespace)
+			if err != nil {
+				log.Error(err, "unable to retrieve pod logs for job "+jobName)
+				return fmt.Errorf("gitops generation job failed")
+			} else {
+				return fmt.Errorf(errMsg)
+			}
+		}
+		time.Sleep(1 * time.Second)
+		select {
+		case <-timeout:
+			stay = false
+		default:
+		}
+	}
+
+	if err != nil && k8sErrors.IsNotFound(err) {
+		return fmt.Errorf("gitops generation job was not found after timeout reached")
+	} else if err != nil {
+		return err
+	}
+
+	return fmt.Errorf("gitops generation job did not complete in time")
+
+}
+
+// Delete job takes in the given gitops job and attempts to delete it
+func DeleteJob(ctx context.Context, client ctrlclient.Client, jobName string, jobNamespace string) error {
+	job := batchv1.Job{
+		TypeMeta: v1.TypeMeta{
+			APIVersion: "batch/v1",
+			Kind:       "Job",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:      jobName,
+			Namespace: jobNamespace,
+		},
+	}
+	return client.Delete(context.Background(), &job, &ctrlclient.DeleteOptions{})
+}
+
+func GetPodLogs(ctx context.Context, client ctrlclient.Client, clientset kubernetes.Interface, jobName string, jobNamespace string) (string, error) {
+	jobPodList, err := clientset.CoreV1().Pods(jobNamespace).List(ctx, v1.ListOptions{LabelSelector: "job-name=" + jobName})
```

Comment:
curious as to why a clientset is being used here, rather than a call with the client, something like:
```
err = client.List(ctx, jobPodList,
		client.InNamespace(jobNamespace),
		client.MatchingLabels{label: "job-name=" + jobName})
```

Reply:
No reason in particular. Figured since I had to use the clientset to get the logs, I might as well be consistent and use it to get the pods too.